### PR TITLE
Provide a helm chart package

### PR DESCRIPTION
**Pull request by babykart**
The process could be something like the following:

1 - new release of garnet
2 - increment the helm chart version (chart version and application version)
3 - generate the README.md using the command `helm-docs charts/garnet`
4 - commit and push to the main branch will trigger the helm-chart github workflow that will package the chart and push it to the helm-charts subdirectory in the Microsoft namespace of the github registry.

From my point of view, there is still a documentation page missing that recalls the commands of the README.md and calls the link to the description of the values.

PS - Feel free to update ...

### CODE REVIEW

**Comment by darrenge** [comment]
FYI - I won't be able to review this until mid next week.

**Comment by darrenge** [comment]
I have not used Helm charts so I am ramping up on what they all entail, so I might be asking basic questions.

Looking at these:
1 - new release of garnet
2 - increment the helm chart version (chart version and application version)

Question:  We wouldn't need to change the "release" of Garnet would we?  https://github.com/microsoft/garnet/blob/main/.azure/pipelines/azure-pipelines-external-release.yml

Also ... where is the helm chart version and application version?  We would have to increment those manually, right?

**Comment by babykart** [comment]
> I have not used Helm charts so I am ramping up on what they all entail, so I might be asking basic questions.
> 
> Looking at these:
> 1 - new release of garnet
> 2 - increment the helm chart version (chart version and application version)
> 
> Question: We wouldn't need to change the "release" of Garnet would we? https://github.com/microsoft/garnet/blob/main/.azure/pipelines/azure-pipelines-external-release.yml

No, indeed.

> 
> Also ... where is the helm chart version and application version? We would have to increment those manually, right?

The helm chart version and the application version (the tag of the docker/oci image) are in the [Chart.yaml](https://github.com/microsoft/garnet/blob/main/charts/garnet/Chart.yaml#L5-L6) files.
You are right: currently these versions have to be incremented manually.
This workflow only packages the helm chart and pushes it into the OCI registry.

**Comment by badrishc** [comment]
Update the comment here to indicate what we need to change as part of every Garnet release:

* https://github.com/microsoft/garnet/blob/main/.azure/pipelines/azure-pipelines-external-release.yml#L3
* https://github.com/microsoft/garnet/blob/main/libs/host/GarnetServer.cs#L52

**Comment by badrishc** [comment]
> Update the comment here to indicate what we need to change as part of every Garnet release:
> 
> * https://github.com/microsoft/garnet/blob/main/.azure/pipelines/azure-pipelines-external-release.yml#L3
> * https://github.com/microsoft/garnet/blob/main/libs/host/GarnetServer.cs#L52

@babykart - ping on this.

**Comment by babykart** [comment]
@badrishc Sorry but I wasn't notified on your previous comment.
It's done.

**Comment by darrenge** [comment]
I saw this comment "update \charts\garnet\Chart.yaml (~line 5 & 6) and generate \charts\garnet\README.md with helm-docs -- 
NOTE".   
1) You don't need the "-- NOTE" on the end
2) Is it Chart.yml or helm-chart.YML?  Note too - it is ".yml" and not ".yaml"
3) Assuming it is helm-chart.Yml --- there is no version number to manually update on line 5 or 6?  If it isn't manually updating a version string, then please specify exactly what is needed.

**Comment by babykart** [comment]
> I saw this comment "update \charts\garnet\Chart.yaml (~line 5 & 6) and generate \charts\garnet\README.md with helm-docs -- NOTE".
> 
>     1. You don't need the "-- NOTE" on the end

Fixed by [683744f](https://github.com/microsoft/garnet/pull/594/commits/683744f7bb20b70afacf50bdb8e5c57e116f70fe)

> 
>     2. Is it Chart.yml or helm-chart.YML?  Note too - it is ".yml" and not ".yaml"

https://github.com/microsoft/garnet/blob/main/charts/garnet/Chart.yaml#L5-L6

> 
>     3. Assuming it is helm-chart.Yml --- there is no version number to manually update on line 5 or 6?  If it isn't manually updating a version string, then please specify exactly what is needed.

https://github.com/microsoft/garnet/blob/main/charts/garnet/Chart.yaml#L5-L6

**Comment by darrenge** [comment]
Thanks for your quick response.  For this part "and generate \charts\garnet\README.md with helm-docs." what does that all entail?  Would the process (as you see it) be:
1) Update string on in .azure/pipelines/azure-pipelines-external-release.yml with latest version of the nuget package (e.g. 1.0.0)
2) Update \libs\host\GarnetServer.cs readonly string version  (~line 45)   -- NOTE - these two values need to be the same 
3) Update \libs\host\GarnetServer.cs readonly string version  (~line 45)   -- NOTE - these two values need to be the same 
4) Update \charts\garnet\Chart.yaml (~line 5 & 6) (just increment by 1 - should be same as #2 and #3 should these numbers be same?)
5) Generate \charts\garnet\README.md with helm-docs.   How?
6) Push all these changes to main
7) Run the release pipeline (azure-pipelines-external-release.yml)

**Comment by babykart** [comment]
On a first iteration and with the aim of having the helm chart process integrated by the teams, I imagined the following process:

1) Keep the actual process with azure pipeline as it,
2) When the tag (ie. v1.0.19) is pushed, it will trigger the github workflows that build and push the docker images (required to the helm chart)
3) Update \charts\garnet\Chart.yaml : the `version` corresponds to the helm chart version and the `appVersion` to the tag of the docker image (1.0.19).
Well, `appVersion` 1.018 -> 1.019, increment the same digit in the helm chart `version` : 0.1.1 -> 0.1.2
4) Generate the \charts\garnet\README.md with the command `helm-docs charts/garnet` (required https://github.com/norwoodj/helm-docs)
5) Commit and push to main branch that will trigger the github workflow (helm package & helm push artifact.)

Actually, the helm chart only implements a standalone deployment but tomorrow we decide to merge a PR that implements the cluster deployment and obviously we want to really release the helm chart (version: 1.0.0):

1) Update \charts\garnet\Chart.yaml: `appVersion` is unchanged but the `version` 0.1.2 will be incremented to the first digit: 1.0.0
2) Generate the \charts\garnet\README.md with the command `helm-docs charts/garnet` (required https://github.com/norwoodj/helm-docs)
3) Commit and push to main branch that will trigger the github workflow (helm package & helm push artifact.)

Once the helm chart is considered by the teams as a finalized release, we could configure a workflow doing 1), 2) & 3) after the successful docker workflows for every new Garnet release.

**Comment by babykart** [comment]
@darrenge Should i consider this PR as abandoned?

**Comment by badrishc** [comment]
> @darrenge Should i consider this PR as abandoned?

We definitely still want this functionality as it appears many people use Helm charts in open source. Do you think this PR does everything necessary for Helm chart support in Garnet (with automated updates when main is updated), or is there more work to be done. There isn't anyone yet on the team with expertise in this so we will have to depend on OSS contributors such as yourself for now. Thanks!

**Comment by darrenge** [comment]
Sorry - I didn't see the question in this PR. I agree with what Badrish is saying.

**Comment by babykart** [comment]
@darrenge  No problem.

@badrishc Yes I do. No more work to have a helm chart package automically when something is commited in `charts/**` directory, even if the generation of the `charts/garnet/README.md` is ommited. After this merge, we just need to test it : I will do the first PR for the example on the last release of Garnet.
[Here](https://github.com/babykart/vault-unseal/blob/main/.github/workflows/release.yml), there is a similar example based on tag.

**Comment by badrishc** [comment]
> @darrenge No problem.
> 
> @badrishc Yes I do. No more work to have a helm chart package automically when something is commited in `charts/**` directory, even if the generation of the `charts/garnet/README.md` is ommited. After this merge, we just need to test it : I will do the first PR for the example on the last release of Garnet. [Here](https://github.com/babykart/vault-unseal/blob/main/.github/workflows/release.yml), there is a similar example based on tag.

Great. We have added a file `Version.props` that tracks the version of Garnet releases. We use it in the release CI to update Garnet server versions so that only one place needs to be updated. There is a powershell script that CI uses to extract the version here: https://github.com/microsoft/garnet/blob/main/.azure/pipelines/azure-pipelines-external-release.yml#L33

Would be great if this PR could be updated to use the same version source for Garnet. Then I think we should be in a position to merge this PR.

**Comment by babykart** [comment]
@darrenge  @badrishc Done for to use `Version.props` in the appVersion of the `charts/garnet/Chart.yaml` file.
Now I think it's a little effort to automatically generate the `charts/garnet/README.md`:

* install helm-docs
* generate `charts/garnet/README.md` file
* Discard changes on `charts/garnet/Chart.yaml` file
* Commit and push to the main branch the generated `charts/garnet/README.md` file

Let me now if you are OK with this.

File: .azure/pipelines/azure-pipelines-external-release.yml
Comment (badrishc) [inline .azure/pipelines/azure-pipelines-external-release.yml:2]: undo the changes to this file, including newline at the end.

File: libs/host/GarnetServer.cs
Comment (badrishc) [inline libs/host/GarnetServer.cs:64]: merge with latest from main to use the new format of version

**Comment by badrishc** [comment]
> Now I think it's a little effort to automatically generate

I am unclear: who will automatically generate and push to main the `charts/garnet/README.md` file? will there be a CI task for this or we have to do it manually?

File: charts/garnet/README.md
Comment (badrishc) [inline charts/garnet/README.md:29]: > "--port","6379","-m","128m","-i","128m"

I dont think these arguments are necessary any more as they default to proper values.

File: charts/garnet/README.md
Comment (badrishc) [inline charts/garnet/README.md:17]: Is there a way users can provide their own `garnet.conf` configuration file for use with the helm chart?

Comment (babykart) [reply charts/garnet/README.md:17]: Actually no in this helm chart but this could be adressed by [this PR](https://github.com/microsoft/garnet/pull/531).

File: .azure/pipelines/azure-pipelines-external-release.yml
Comment (babykart) [reply .azure/pipelines/azure-pipelines-external-release.yml]: Fixed by 3b838b4fd9fcfe785ceeb9a6ae52a36c4ddbfcb5

File: libs/host/GarnetServer.cs
Comment (babykart) [reply libs/host/GarnetServer.cs]: Fixed by 3b838b4fd9fcfe785ceeb9a6ae52a36c4ddbfcb5

File: charts/garnet/README.md
Comment (babykart) [reply charts/garnet/README.md:29]: Fixed by 1cd1e40c4b3f0111b27724dcde8b68c25dfc9cf0

File: charts/garnet/README.md
Comment (badrishc) [reply charts/garnet/README.md:17]: OK would be great to see how to port this into your helm chart in a separate PR, as the other PR seems to create a duplicate.

**Comment by badrishc** [comment]
> On a first iteration and with the aim of having the helm chart process integrated by the teams, I imagined the following process:
> 
> 1. Keep the actual process with azure pipeline as it,
> 2. When the tag (ie. v1.0.19) is pushed, it will trigger the github workflows that build and push the docker images (required for the helm chart)
> 3. Update \charts\garnet\Chart.yaml : the `version` corresponds to the helm chart version and the `appVersion` to the tag of the docker image (1.0.19).
>    Well, `appVersion` 1.018 -> 1.019, increment the same digit in the helm chart `version` : 0.1.1 -> 0.1.2
> 4. Generate the \charts\garnet\README.md with the command `helm-docs charts/garnet` (required https://github.com/norwoodj/helm-docs)
> 5. Commit and push to the main branch that will trigger the github workflow (helm package & helm push artifact.)
> 
> Actually, the helm chart only implements a standalone deployment but tomorrow we decide to merge a PR that implements the cluster deployment and obviously we want to really release the helm chart (version: 1.0.0):
> 
> 1. Update \charts\garnet\Chart.yaml: `appVersion` is unchanged but the `version` 0.1.2 will be incremented to the first digit: 1.0.0
> 2. Generate the \charts\garnet\README.md with the command `helm-docs charts/garnet` (required https://github.com/norwoodj/helm-docs)
> 3. Commit and push to the main branch that will trigger the github workflow (helm package & helm push artifact.)
> 
> Once the helm chart is considered by the teams as a finalized release, we could configure a workflow doing 1), 2) & 3) after the successful docker workflows for every new Garnet release.

could you edit this process flow to reflect the latest process after these changes?

**Review by badrishc — APPROVED** [review]

**Comment by badrishc** [comment]
Approved. I think we are ready to merge this, once you give the go-ahead that your commits are complete for this PR.

**Comment by babykart** [comment]
> > Now I think it's a little effort to automatically generate
> 
> I am unclear: who will automatically generate and push to main the `charts/garnet/README.md` file? will there be a CI task for this or we have to do it manually?

Sorry, in this state, it's a manual operation but my proposition is to implement these steps in this workflow.

* install helm-docs
* generate `charts/garnet/README.md` file
* Discard changes on `charts/garnet/Chart.yaml` file
* Commit and push to the main branch the generated `charts/garnet/README.md` file

As it, a future helm chart will be built only with one human commit incrementing the version in `charts/garnet/Chart.yaml`.

**Comment by badrishc** [comment]
> > > Now I think it's a little effort to automatically generate
> > 
> > 
> > I am unclear: who will automatically generate and push to main the `charts/garnet/README.md` file? will there be a CI task for this or we have to do it manually?
> 
> Sorry, in this state, it's a manual operation but my proposition is to implement these steps in this workflow.
> 
> * install helm-docs
> * generate `charts/garnet/README.md` file
> * Discard changes on `charts/garnet/Chart.yaml` file
> * Commit and push to the main branch the generated `charts/garnet/README.md` file
> 
> As it, a future helm chart will be built only with one human commit incrementing the version in `charts/garnet/Chart.yaml`.

Sounds good, after this PR merges, you will create a PR for that first manual commit to README?

**Comment by babykart** [comment]
Yes it seems good to me to finish this PR (just let me the time , probably tomorrow, to edit the workflow in the head of the comment as you mentioned)
In the next days, then I will open another one with the manual process of incrementing the version of the `Chart.yaml` file and the generation of the `README.md` file
Once the build is validated, another PR to automate the generation of the `README.md` and its commit.
In parallel, I could work on the functionality of using a configuration file.

**Pull request merged**
